backends/oauth2.go

```diff
@@ -0,0 +1,361 @@
+		cache.writeTopics = info.MQTT.Topics.Write
+		cache.subscribeTopics = info.MQTT.Topics.Subscribe
+		cache.denyTopics = info.MQTT.Topics.Deny
+		cache.updatedAt = time.Now()
+	}
+
+	log.Debugf("  user is superuser: %t", cache.superuser)
+	log.Debugf("  topics with read permission: %s", cache.readTopics)
+	log.Debugf("  topics with write permission: %s", cache.writeTopics)
+	log.Debugf("  topics with subscribe permission: %s", cache.subscribeTopics)
+	log.Debugf("  denied topics: %s", cache.denyTopics)
+
+	return nil
+}
+
+func (o *Oauth2) createUserWithClientCredentials(username, password, clientid string) (bool, error) {
+	clientcredentialsConfig := go_clientcredentials.Config{
+		ClientID:     username,
+		ClientSecret: password,
+		TokenURL:     o.tokenUrl,
+		Scopes:       o.scopesSplit,
+	}
+
+	clientcredentialsClient := clientcredentialsConfig.Client(context.Background())
+
+	o.userCache[username] = userState{
+		username:  username,
+		superuser: false,
+		createdAt: time.Now(),
+		updatedAt: time.Unix(0, 0),
+		client:    clientcredentialsClient,
+	}
+
+	cache, _ := o.userCache[username]
+	err := o.updateCache(&cache)
+	if err != nil {
+		return false, err
+	}
+
+	return true, nil
+}
+
+func (o *Oauth2) createUserWithToken(accessToken, clientid string) (bool, error) {
+	token := &go_oauth2.Token{
+		AccessToken: accessToken,
+		TokenType:   "Bearer",
+	}
+
+	oauth2Config := go_oauth2.Config{
+		Scopes:      o.scopesSplit,
+		RedirectURL: "",
+		Endpoint: go_oauth2.Endpoint{
+			TokenURL: o.tokenUrl,
+			AuthURL:  "",
+		},
+	}
+
+	client := oauth2Config.Client(context.Background(), token)
+
+	o.userCache[accessToken] = userState{
+		username:  accessToken,
+		createdAt: time.Now(),
+		updatedAt: time.Unix(0, 0),
+		client:    client,
+	}
+
+	cache, _ := o.userCache[accessToken]
+	err := o.updateCache(&cache)
+	if err != nil {
+		return false, err
+	}
+
+	return true, nil
+}
+
+func (o *Oauth2) getUserInfo(client *http.Client) (*UserInfo, error) {
+	req, _ := http.NewRequest("GET", o.userInfoURL, nil)
+	log.Debugf("Performing request: %s %s://%s%s", req.Method, req.URL.Scheme, req.URL.Host, req.URL.Path)
+	if log.IsLevelEnabled(log.DebugLevel) {
+		token, err := client.Transport.(*go_oauth2.Transport).Source.Token()
+		if err != nil {
+			log.Debug(err)
+			return nil, err
+		}
+		log.Debugf("Using Token: %s", token.AccessToken)
+	}
+	resp, err := client.Do(req)
+	if err != nil {
+		log.Debug(err)
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	bodyBytes, err := io.ReadAll(resp.Body)
+	if err != nil {
+		log.Fatal(err)
+	}
+	bodyString := string(bodyBytes)
+	log.Debugf("Response: %s", bodyString)
+
+	info := UserInfo{}
+
+	err = json.Unmarshal([]byte(bodyString), &info)
+	if err != nil {
+		return nil, err
+	}
+
+	return &info, nil
+}
+
+func (o *Oauth2) checkAccessToTopic(topic string, acc int32, cache *userState, username string, clientid string) bool {
+	log.Debugf("Check for acl level %d", acc)
+
+	if acc == MOSQ_ACL_NONE {
```

Comment:
Could we do a switch over the possible `acc` value instead of multiple `if` and `return`?